File: deno/lib/validators/regex.ts
import {
  ServiceException,
} from "/lib/exceptions/service.exception.ts";

import {
  ns,
} from "/src/base/i18n/i18n.ts";

export async function regex(
  value?: string | null,
  reg?: string | null,
  label?: string,
) {
  if (value == null) {
    return;
  }
  if (reg == null) {
    return;
  }
  const regex = new RegExp(reg, "gm");
  if (regex.test(value)) {
    return;
  }
  label = label || "";
  const err_msg = await ns(
    "格式不正确",
  );
  throw new ServiceException(`${ label } ${ err_msg }`, "regex");
}

File: rust/src/common/validators/regex.rs
use anyhow::Result;
use regex::Regex;

use crate::common::context::Ctx;
use crate::common::context::SrvErr;
use crate::src::base::i18n::i18n_dao::ns;

#[allow(dead_code)]
pub async fn regex(
  ctx: &mut Ctx,
  value: Option<String>,
  regex_str: &'static str,
  label: &str,
) -> Result<()> {
  
  if value.is_none() {
    return Ok(());
  }
  let value = value.unwrap();
  let regex = Regex::new(regex_str)?;
  if regex.is_match(&value) {
    return Ok(());
  }
  
  let msg = ns(
    ctx,
    "格式不正确".to_owned(),
    None,
  ).await?;
  
  let mut err_msg = String::new();
  err_msg.push_str(label);
  err_msg.push(' ');
  err_msg.push_str(&msg);
  let err_msg = err_msg;
  
  Err(SrvErr::msg(err_msg).into())
}
